SVT:Cannot find component for expression

My Jakarta EE10 application is failing after enabling Jakarta EE11 features in the server.xml file with the following exception.
Looks like Ajax calls are failing in one of the JSF facelet page. I also captured the logs with this tracespec "*=info:com.ibm.ws.webcontainer*=all:com.ibm.wsspi.webcontainer*=all:com.ibm.ws.webcontainer31*=all:HTTPChannel=all:TCPChannel=all:websockets=all:com.ibm.ws.jsf*=all:org.apache.myfaces*=all"

**Logs**
[logs.zip](https://github.com/user-attachments/files/17134082/logs.zip)

**server.xml** 

```
    <featureManager>
		<feature>xmlWS-4.0</feature>
		<feature>xmlBinding-4.0</feature>
		<!-- <feature>managedBeans-2.0</feature> -->
		<feature>appSecurity-6.0</feature>
        <feature>jdbc-4.3</feature>
    	<feature>concurrent-3.1</feature>
		<!-- <feature>osgiConsole-1.0</feature> -->
		 <feature>usr:WASPersonaWebServicesHandlerFeature</feature>
		<feature>websocket-2.2</feature>
    	<feature>expressionLanguage-6.0</feature>
		<feature>validation-3.1</feature>
		<feature>cdi-4.1</feature>
		<feature>faces-4.1</feature>
		<feature>enterpriseBeansLite-4.0</feature>
		<feature>persistence-3.2</feature>
		<feature>restfulWS-4.0</feature>
 		<!--  <feature>wsAtomicTransaction-1.2</feature>--> 		
    	<feature>mpHealth-4.0</feature>
    	<feature>mpMetrics-5.1</feature>
    	<feature>mpConfig-3.1</feature>
    	<feature>mpTelemetry-2.0</feature>
		<!-- Features with the versions -->
    	<feature>jndi-1.0</feature>
		<feature>localConnector-1.0</feature>
		<feature>monitor-1.0</feature>
		<feature>ssl-1.0</feature>
		<feature>sessionDatabase-1.0</feature>
    	<feature>json-1.0</feature>
	</featureManager>

```

```
[err] jakarta.faces.component.search.ComponentNotFoundException: Cannot find component for expression "cartForm:cartTable cartForm:subTotal cartForm:taxAmount cartForm:total" referenced from "shoppingCartForm_form:updateButton".
[err] 	at org.apache.myfaces.component.search.SearchExpressionHandlerImpl.resolveClientIds(SearchExpressionHandlerImpl.java:179)
[err] 	at org.apache.myfaces.renderkit.html.util.AjaxScriptBuilder.appendIds(AjaxScriptBuilder.java:278)
[err] 	at org.apache.myfaces.renderkit.html.util.AjaxScriptBuilder.build(AjaxScriptBuilder.java:205)
[err] 	at org.apache.myfaces.renderkit.html.util.AjaxScriptBuilder.build(AjaxScriptBuilder.java:117)
[err] 	at org.apache.myfaces.renderkit.html.HtmlAjaxBehaviorRenderer.getScript(HtmlAjaxBehaviorRenderer.java:71)
[err] 	at jakarta.faces.component.behavior.ClientBehaviorBase.getScript(ClientBehaviorBase.java:92)
[err] 	at org.apache.myfaces.renderkit.html.util.ClientBehaviorRendererUtils.appendClientBehaviourScript(ClientBehaviorRendererUtils.java:208)
[err] 	at org.apache.myfaces.renderkit.html.util.ClientBehaviorRendererUtils.getClientBehaviorScript(ClientBehaviorRendererUtils.java:186)
[err] 	at org.apache.myfaces.renderkit.html.util.ClientBehaviorRendererUtils.buildBehaviorChain(ClientBehaviorRendererUtils.java:349)
[err] 	at org.apache.myfaces.renderkit.html.util.ClientBehaviorRendererUtils.buildBehaviorChain(ClientBehaviorRendererUtils.java:315)
[err] 	at org.apache.myfaces.renderkit.html.base.HtmlButtonRendererBase.buildBehaviorizedOnClick(HtmlButtonRendererBase.java:340)
[err] 	at org.apache.myfaces.renderkit.html.base.HtmlButtonRendererBase.encodeBegin(HtmlButtonRendererBase.java:186)
[err] 	at jakarta.faces.component.UIComponentBase.encodeBegin(UIComponentBase.java:561)
[err] 	at jakarta.faces.component.UIComponentBase.encodeAll(UIComponentBase.java:495)
[err] 	at jakarta.faces.component.UIComponentBase.encodeAll(UIComponentBase.java:519)
[err] 	at org.apache.myfaces.renderkit.RendererUtils.renderChildren(RendererUtils.java:536)
[err] 	at org.apache.myfaces.renderkit.html.base.HtmlGroupRendererBase.encodeEnd(HtmlGroupRendererBase.java:154)
[err] 	at jakarta.faces.component.UIComponentBase.encodeEnd(UIComponentBase.java:634)
[err] 	at jakarta.faces.component.UIComponentBase.encodeAll(UIComponentBase.java:523)
[err] 	at jakarta.faces.component.UIComponentBase.encodeAll(UIComponentBase.java:519)
[err] 	at jakarta.faces.component.UIComponentBase.encodeAll(UIComponentBase.java:519)
[err] 	at org.apache.myfaces.view.facelets.FaceletViewDeclarationLanguage.renderView(FaceletViewDeclarationLanguage.java:1779)
[err] 	at [internal classes]
[err] 	at jakarta.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:74)
[err] 	at org.apache.myfaces.lifecycle.RenderResponseExecutor.execute(RenderResponseExecutor.java:122)
[err] 	at [internal classes]
[err] 	at jakarta.faces.webapp.FacesServlet.service(FacesServlet.java:225)
[err] 	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1266)
[err] 	at [internal classes]
[err] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
[err] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
[err] 	at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR   ] SRVE0777E: Exception thrown by application class 'io.openliberty.webcontainer61.osgi.webapp.WebAppDispatcherContext61.sendRedirect:81'
java.lang.IllegalStateException: CWWWC0007E: The response sendRedirect is not allowed after the response is committed.
	at io.openliberty.webcontainer61.osgi.webapp.WebAppDispatcherContext61.sendRedirect(WebAppDispatcherContext61.java:81)
	at [internal classes]
	at com.ibm.websphere.svt.gs.gsjsfweb.utils.GSJSFExceptionHandler.handle(GSJSFExceptionHandler.java:75)
	at org.apache.myfaces.lifecycle.LifecycleImpl.render(LifecycleImpl.java:261)
	at [internal classes]

[ERROR   ] SRVE0315E: An exception occurred: java.lang.Throwable: java.lang.IllegalStateException: CWWWC0007E: The response sendRedirect is not allowed after the response is committed.
	at com.ibm.ws.webcontainer.webapp.WebApp.handleRequest(WebApp.java:5154)
	at [internal classes]
Caused by: java.lang.IllegalStateException: CWWWC0007E: The response sendRedirect is not allowed after the response is committed.
	at io.openliberty.webcontainer61.osgi.webapp.WebAppDispatcherContext61.sendRedirect(WebAppDispatcherContext61.java:81)
	at [internal classes]
	at com.ibm.websphere.svt.gs.gsjsfweb.utils.GSJSFExceptionHandler.handle(GSJSFExceptionHandler.java:75)
	at org.apache.myfaces.lifecycle.LifecycleImpl.render(LifecycleImpl.java:261)
	... 1 more
```

--- a/dev/com.ibm.ws.jsf.2.2_fat.2/bnd.bnd
+++ b/dev/com.ibm.ws.jsf.2.2_fat.2/bnd.bnd
@@ -48,7 +48,8 @@ src: \
     test-applications/FunctionMapper.war/src,\
     test-applications/ViewScopeLeak.war/src,\
     test-applications/MYFACES-4512.war/src,\
-    test-applications/PH55398.war/src
+    test-applications/PH55398.war/src,\
+    test-applications/MYFACES-4695.war/src
 
 fat.project: true
 

--- a/dev/com.ibm.ws.jsf.2.2_fat.2/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22APARSeleniumTests.java
+++ b/dev/com.ibm.ws.jsf.2.2_fat.2/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22APARSeleniumTests.java
@@ -10,6 +10,9 @@
 package com.ibm.ws.jsf22.fat.tests;
 
 import static org.junit.Assert.assertEquals;
+import static componenttest.annotation.SkipForRepeat.EE10_FEATURES;
+import static componenttest.annotation.SkipForRepeat.EE9_FEATURES;
+import static componenttest.annotation.SkipForRepeat.EE8_FEATURES;
 import static org.junit.Assert.assertTrue;
 
 import org.junit.After;
@@ -36,6 +39,7 @@ import io.openliberty.faces.fat.selenium.util.internal.ExtendedWebDriver;
 import io.openliberty.faces.fat.selenium.util.internal.WebPage;
 
 import componenttest.annotation.Server;
+import componenttest.annotation.SkipForRepeat;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.custom.junit.runner.Mode;
 import componenttest.custom.junit.runner.Mode.TestMode;
@@ -66,6 +70,8 @@ public class JSF22APARSeleniumTests {
     public static void setup() throws Exception {
 
         ShrinkHelper.defaultDropinApp(jsf22APARSeleniumServer, "PH55398.war", "com.ibm.ws.jsf22.fat.PH55398.bean");
+
+        ShrinkHelper.defaultDropinApp(jsf22APARSeleniumServer, "MYFACES-4695.war", "com.ibm.ws.jsf22.fat.myfaces4695");
 
         jsf22APARSeleniumServer.startServer(c.getSimpleName() + ".log");
 
@@ -162,5 +168,23 @@ public class JSF22APARSeleniumTests {
             assertEquals("false", page.findElement(By.id("form1:output")).getText());
 
     }
-    
+
+    // This problem was found by SVT where a nested element in another naming container
+    // is not updated at all.  Our test expects the "cartForm:result" to be updated.
+    @SkipForRepeat({ EE10_FEATURES, EE9_FEATURES, EE8_FEATURES }) // skipped until new releases are pulled in. 
+    @Test
+    public void testMyFaces4695() throws Exception {
+        String url = JSFUtils.createSeleniumURLString(jsf22APARSeleniumServer, "MYFACES-4695", "index.xhtml");
+        WebPage page = new WebPage(driver);
+
+        page.get(url);
+        page.waitForPageToLoad();
+
+        WebElement submitButton = page.findElement(By.id("shoppingCartForm:submit"));
+        submitButton.click(); //invoke ajax to re-render the cartForm:result element
+        page.waitReqJs();
+
+        assertEquals("success", page.findElement(By.id("cartForm:result")).getText());
+    }
+        
 }

--- a/dev/com.ibm.ws.jsf.2.2_fat.2/test-applications/MYFACES-4695.war/resources/WEB-INF/web.xml
+++ b/dev/com.ibm.ws.jsf.2.2_fat.2/test-applications/MYFACES-4695.war/resources/WEB-INF/web.xml
@@ -1,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2024 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<web-app id="WebApp_ID" version="3.1" 
+    xmlns="http://xmlns.jcp.org/xml/ns/javaee" 
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
+    xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd">
+	
+  <display-name>MYFACES-4695</display-name>
+  
+  <servlet>
+      <servlet-name>Faces Servlet</servlet-name>
+      <servlet-class>javax.faces.webapp.FacesServlet</servlet-class>
+  </servlet>
+  
+  <servlet-mapping>
+      <servlet-name>Faces Servlet</servlet-name>
+      <url-pattern>*.xhtml</url-pattern>
+  </servlet-mapping>
+
+  <!-- Welcome page -->
+  <welcome-file-list>
+      <welcome-file>index.xhtml</welcome-file>
+  </welcome-file-list>
+    
+</web-app>

--- a/dev/com.ibm.ws.jsf.2.2_fat.2/test-applications/MYFACES-4695.war/resources/index.xhtml
+++ b/dev/com.ibm.ws.jsf.2.2_fat.2/test-applications/MYFACES-4695.war/resources/index.xhtml
@@ -1,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2024 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml" xmlns:h="http://xmlns.jcp.org/jsf/html"
+    xmlns:f="http://xmlns.jcp.org/jsf/core">
+    <head>
+        <title>MYFACES-4512</title>
+    </head>
+
+    <body>
+        <h:form id="cartForm">
+            <h:outputText id="result" value="#{bean.result}" />
+        </h:form>
+        <h:form id="shoppingCartForm" name="shoppingCartForm">
+            <h:commandButton type="submit" value="Update" action="#{bean.click}" id="submit">
+                <f:ajax execute="@form" render="cartForm:result" />
+            </h:commandButton>
+        </h:form>
+    </body>
+</html>

--- a/dev/com.ibm.ws.jsf.2.2_fat.2/test-applications/MYFACES-4695.war/src/com/ibm/ws/jsf22/fat/myfaces4695/MyFaces4695Bean.java
+++ b/dev/com.ibm.ws.jsf.2.2_fat.2/test-applications/MYFACES-4695.war/src/com/ibm/ws/jsf22/fat/myfaces4695/MyFaces4695Bean.java
@@ -1,0 +1,30 @@
+/*******************************************************************************
+ * Copyright (c) 2024 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.jsf22.fat.myfaces4695;
+
+import java.io.Serializable;
+
+import javax.enterprise.context.SessionScoped;
+import javax.inject.Named;
+
+@SessionScoped
+@Named("bean") // for simplicity 
+public class MyFaces4695Bean implements Serializable {
+
+    private String result;
+
+    public String getResult() {
+        return result;
+    }
+
+    public void click() {
+        result = "success";
+    }
+}

--- a/dev/io.openliberty.org.apache.myfaces.4.1/src/org/apache/myfaces/component/search/SearchExpressionHandlerImpl.java
+++ b/dev/io.openliberty.org.apache.myfaces.4.1/src/org/apache/myfaces/component/search/SearchExpressionHandlerImpl.java
@@ -1,0 +1,864 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.myfaces.component.search;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import jakarta.faces.FacesException;
+import jakarta.faces.component.ContextCallback;
+import jakarta.faces.component.NamingContainer;
+import jakarta.faces.component.UIComponent;
+import jakarta.faces.component.UIViewRoot;
+import jakarta.faces.component.search.ComponentNotFoundException;
+import jakarta.faces.component.search.SearchExpressionContext;
+import jakarta.faces.component.search.SearchExpressionHandler;
+import jakarta.faces.component.search.SearchExpressionHint;
+import jakarta.faces.component.search.SearchKeywordContext;
+import jakarta.faces.context.FacesContext;
+import org.apache.myfaces.core.api.shared.ComponentUtils;
+import org.apache.myfaces.core.api.shared.lang.SharedStringBuilder;
+import org.apache.myfaces.util.lang.StringUtils;
+
+/**
+ *
+ */
+public class SearchExpressionHandlerImpl extends SearchExpressionHandler
+{
+    private static final String SB_SPLIT = SearchExpressionHandlerImpl.class.getName() + "#split";
+
+    protected void addHint(SearchExpressionContext searchExpressionContext, SearchExpressionHint hint)
+    {
+        // already available
+        searchExpressionContext.getExpressionHints().add(hint);
+    }
+    
+    protected boolean isHintSet(SearchExpressionContext searchExpressionContext, SearchExpressionHint hint)
+    {
+        if (searchExpressionContext.getExpressionHints() == null)
+        {
+            return false;
+        }
+        
+        return searchExpressionContext.getExpressionHints().contains(hint);
+    }
+    
+    @Override
+    public String resolveClientId(SearchExpressionContext searchExpressionContext, String expression)
+    {
+        if (expression == null)
+        {
+            expression = "";
+        }
+        else
+        {
+            expression = expression.trim();
+        }
+        
+        FacesContext facesContext = searchExpressionContext.getFacesContext();
+        SearchExpressionHandler handler = facesContext.getApplication().getSearchExpressionHandler();
+        
+        addHint(searchExpressionContext, SearchExpressionHint.RESOLVE_SINGLE_COMPONENT);
+
+        if (handler.isPassthroughExpression(searchExpressionContext, expression))
+        {
+            return expression;
+        }
+        else
+        {
+            CollectClientIdCallback callback = null;
+            if (!expression.isEmpty())
+            {
+                callback = new CollectClientIdCallback();
+                handler.invokeOnComponent(
+                        searchExpressionContext, searchExpressionContext.getSource(), expression, callback);
+            }
+
+            if (callback == null || !callback.isClientIdFound())
+            {
+                if (isHintSet(searchExpressionContext, SearchExpressionHint.IGNORE_NO_RESULT))
+                {
+                    return null;
+                }
+                else
+                {
+                    throw new ComponentNotFoundException("Cannot find component for expression \""
+                        + expression + "\" referenced from \""
+                        + searchExpressionContext.getSource().getClientId(facesContext) + "\".");
+                }
+            }
+            return callback.getClientId();
+        }
+    }
+
+    private static class CollectClientIdCallback implements ContextCallback
+    {
+        private String clientId = null;
+
+        @Override
+        public void invokeContextCallback(FacesContext context, UIComponent target)
+        {
+            if (clientId == null)
+            {
+                clientId = target.getClientId(context);
+            }
+        }
+
+        private String getClientId()
+        {
+            return clientId;
+        }
+
+        private boolean isClientIdFound()
+        {
+            return clientId != null;
+        }
+    }
+
+    @Override
+    public List<String> resolveClientIds(SearchExpressionContext searchExpressionContext, String expressions)
+    {
+        if (expressions == null)
+        {
+            expressions = "";
+        }
+        else
+        {
+            expressions = expressions.trim();
+        }
+        
+        FacesContext facesContext = searchExpressionContext.getFacesContext();
+        SearchExpressionHandler handler = facesContext.getApplication().getSearchExpressionHandler();
+
+        CollectClientIdsCallback callback = null;
+        if (!expressions.isEmpty())
+        {
+            callback = new CollectClientIdsCallback();
+            for (String expression : handler.splitExpressions(facesContext, expressions))
+            {
+                if (handler.isPassthroughExpression(searchExpressionContext, expression))
+                {
+                    // It will be resolved in the client, just add the expression.
+                    callback.addClientId(expression);
+                }
+                else
+                {
+                    handler.invokeOnComponent(
+                            searchExpressionContext, searchExpressionContext.getSource(), expression, callback);
+                }
+            }
+        }
+
+        if (callback == null || !callback.isClientIdFound())
+        {
+            if (isHintSet(searchExpressionContext, SearchExpressionHint.IGNORE_NO_RESULT))
+            {
+                return Collections.emptyList();
+            }
+            else
+            {
+                throw new ComponentNotFoundException("Cannot find component for expression \""
+                    + expressions + "\" referenced from \""
+                    + searchExpressionContext.getSource().getClientId(facesContext) + "\".");
+            }
+        }
+        return callback.getClientIds();
+    }
+
+    private static class CollectClientIdsCallback implements ContextCallback
+    {
+        private List<String> clientIds = null;
+
+        @Override
+        public void invokeContextCallback(FacesContext context, UIComponent target)
+        {
+            if (clientIds == null)
+            {
+                clientIds = new ArrayList<>(5);
+            }
+            clientIds.add(target.getClientId(context));
+        }
+
+        private List<String> getClientIds()
+        {
+            return clientIds == null ? Collections.emptyList() : clientIds;
+        }
+
+        private boolean isClientIdFound()
+        {
+            return clientIds != null;
+        }
+
+        private void addClientId(String clientId)
+        {
+            if (clientIds == null)
+            {
+                clientIds = new ArrayList<>(5);
+            }
+            clientIds.add(clientId);
+        }
+    }
+
+    @Override
+    public void resolveComponent(SearchExpressionContext searchExpressionContext, String expression,
+        ContextCallback callback)
+    {
+        if (expression == null)
+        {
+            expression = "";
+        }
+        else
+        {
+            expression = expression.trim();
+        }
+        
+        FacesContext facesContext = searchExpressionContext.getFacesContext();
+        SearchExpressionHandler handler = facesContext.getApplication().getSearchExpressionHandler();
+
+        addHint(searchExpressionContext, SearchExpressionHint.RESOLVE_SINGLE_COMPONENT);
+
+        SingleInvocationCallback checkCallback = null;
+        if (!expression.isEmpty())
+        {
+            checkCallback = new SingleInvocationCallback(callback);
+            handler.invokeOnComponent(searchExpressionContext, searchExpressionContext.getSource(),
+                    expression, checkCallback);
+        }
+
+        if (checkCallback == null || !checkCallback.isInvoked())
+        {
+            if (isHintSet(searchExpressionContext, SearchExpressionHint.IGNORE_NO_RESULT))
+            {
+                //Ignore
+            }
+            else
+            {
+                throw new ComponentNotFoundException("Cannot find component for expression \""
+                    + expression + "\" referenced from \""
+                    + searchExpressionContext.getSource().getClientId(facesContext) + "\".");
+            }
+        }
+    }
+
+    private static class SingleInvocationCallback implements ContextCallback
+    {
+        private boolean invoked;
+
+        private final ContextCallback innerCallback;
+
+        public SingleInvocationCallback(ContextCallback innerCallback)
+        {
+            this.innerCallback = innerCallback;
+            this.invoked = false;
+        }
+
+        @Override
+        public void invokeContextCallback(FacesContext context, UIComponent target)
+        {
+            if (!isInvoked())
+            {
+                try
+                {
+                    innerCallback.invokeContextCallback(context, target);
+                }
+                finally
+                {
+                    invoked = true;
+                }
+            }
+        }
+
+        public boolean isInvoked()
+        {
+            return invoked;
+        }
+    }
+
+    @Override
+    public void resolveComponents(SearchExpressionContext searchExpressionContext, String expressions,
+            ContextCallback callback)
+    {
+        if (expressions == null)
+        {
+            expressions = "";
+        }
+        else
+        {
+            expressions = expressions.trim();
+        }
+        
+        FacesContext facesContext = searchExpressionContext.getFacesContext();
+        SearchExpressionHandler handler = facesContext.getApplication().getSearchExpressionHandler();
+        
+        MultipleInvocationCallback checkCallback = null;
+
+        if (!expressions.isEmpty())
+        {
+            checkCallback = new MultipleInvocationCallback(callback);
+            for (String expression : handler.splitExpressions(facesContext, expressions))
+            {
+                handler.invokeOnComponent(searchExpressionContext, searchExpressionContext.getSource(),
+                        expression, checkCallback);
+            }
+        }
+
+        if (checkCallback == null || !checkCallback.isInvoked())
+        {
+            if (isHintSet(searchExpressionContext, SearchExpressionHint.IGNORE_NO_RESULT))
+            {
+                //Ignore
+            }
+            else
+            {
+                throw new ComponentNotFoundException("Cannot find component for expression \""
+                    + expressions + "\" referenced from \""
+                    + searchExpressionContext.getSource().getClientId(facesContext) + "\".");
+            }
+        }
+    }
+
+    private static class MultipleInvocationCallback implements ContextCallback
+    {
+        private boolean invoked;
+
+        private final ContextCallback innerCallback;
+
+        public MultipleInvocationCallback(ContextCallback innerCallback)
+        {
+            this.innerCallback = innerCallback;
+            this.invoked = false;
+        }
+
+        @Override
+        public void invokeContextCallback(FacesContext context, UIComponent target)
+        {
+            try
+            {
+                innerCallback.invokeContextCallback(context, target);
+            }
+            finally
+            {
+                invoked = true;
+            }
+        }
+
+        public boolean isInvoked()
+        {
+            return invoked;
+        }
+    }
+
+    @Override
+    public void invokeOnComponent(final SearchExpressionContext searchExpressionContext,
+            UIComponent previous, String topExpression, final ContextCallback topCallback)
+    {
+        if (topExpression == null)
+        {
+            topExpression = "";
+        }
+        else
+        {
+            topExpression = topExpression.trim();
+        }
+        
+        // Command pattern to apply the keyword or command to the base and then invoke the callback
+        FacesContext facesContext = searchExpressionContext.getFacesContext();
+
+        SearchExpressionHandler handler = facesContext.getApplication().getSearchExpressionHandler();
+        
+        //Step 1: find base
+        //  Case ':' (root)
+        char separatorChar = facesContext.getNamingContainerSeparatorChar();
+        if (topExpression.charAt(0) == separatorChar)
+        {
+            UIComponent findBase = ComponentUtils.findRootComponent(previous);
+            handler.invokeOnComponent(searchExpressionContext, findBase, topExpression.substring(1), topCallback);
+            return;
+        }
+
+        //Step 2: Once you have a base where you can start, apply an expression
+        if (topExpression.charAt(0) == KEYWORD_PREFIX.charAt(0))
+        {
+            // A keyword means apply a command over the current source using an expression and the result must be
+            // feedback into the algorithm.
+
+            String command = extractKeyword(topExpression, 1, separatorChar);
+            final String remaining =
+                    command.length()+1 < topExpression.length() ?
+                        topExpression.substring(1+command.length()+1) : null;
+
+            // If the keyword is @child, @composite, @form, @namingcontainer, @next, @none, @parent, @previous,
+            // @root, @this ,  all commands change the source to be applied the action
+            if (remaining != null)
+            {
+                if (facesContext.getApplication().getSearchKeywordResolver().isLeaf(searchExpressionContext, command))
+                {
+                    throw new FacesException("Expression cannot have keywords or ids at the right side: "+command);
+                }
+
+                this.applyKeyword(searchExpressionContext, previous, command, remaining, new ContextCallback()
+                    {
+                        @Override
+                        public void invokeContextCallback(FacesContext facesContext, UIComponent target)
+                        {
+                            handler.invokeOnComponent(
+                                    searchExpressionContext, target, remaining, topCallback);
+                        }
+                    });
+            }
+            else
+            {
+                // Command completed, apply parent callback
+                this.applyKeyword(searchExpressionContext, previous, command, null, topCallback);
+            }
+            
+        }
+        else
+        {
+
+            //Split expression into tokens and apply loop
+            String nextExpression = null;
+            String expression;
+            if (topExpression.indexOf(":@") > 0)
+            {
+                int idx = topExpression.indexOf(":@");
+                nextExpression = topExpression.substring(idx+1);
+                expression = topExpression.substring(0, idx);
+            }
+            else
+            {
+                expression = topExpression;
+            }
+
+            // Use findComponent(...) passing the expression provided
+            UIComponent target = previous.findComponent(expression);
+            if (target == null)
+            {
+                // If no component is found ...
+                // First try to find the base component.
+
+                // Extract the base id from the expression string
+                int idx = expression.indexOf(separatorChar);
+                String base = idx > 0 ? expression.substring(0, idx) : expression;
+
+                // From the context component clientId, check if the base is part of the clientId
+                String contextClientId = previous.getClientId(facesContext);
+                int startCommon = contextClientId.lastIndexOf(base+facesContext.getNamingContainerSeparatorChar());
+                if (startCommon >= 0
+                    && (startCommon == 0 || contextClientId.charAt(startCommon-1) == separatorChar )
+                    && (startCommon+base.length() <= contextClientId.length()-1 ||
+                        contextClientId.charAt(startCommon+base.length()+1) == separatorChar ))
+                {
+                    // If there is a match, try to find a the first parent component whose id is equals to
+                    // the base id
+                    UIComponent parent = previous;
+                    while (parent != null )
+                    {
+                        if (base.equals(parent.getId()) && parent instanceof NamingContainer)
+                        {
+                            break;
+                        }
+                        else
+                        {
+                            parent = parent.getParent();
+                        }
+                    }
+
+                    // if a base component is found ...
+                    if (parent != null)
+                    {
+                        target = parent.findComponent(expression);
+                        if (target == null && !searchExpressionContext.getExpressionHints().contains(
+                                SearchExpressionHint.SKIP_VIRTUAL_COMPONENTS))
+                        {
+                            contextClientId = parent.getClientId(facesContext);
+                            // If no component is found,
+                            String targetClientId = contextClientId.substring(0, startCommon+base.length()) +
+                                    expression.substring(base.length());
+
+                            if (nextExpression != null)
+                            {
+                                final String childExpression = nextExpression;
+
+                                parent.invokeOnComponent(facesContext, targetClientId, new ContextCallback() 
+                                {
+                                    @Override
+                                    public void invokeContextCallback(FacesContext context, UIComponent target)
+                                    {
+                                        handler.invokeOnComponent(
+                                                searchExpressionContext, target, childExpression, topCallback);
+                                    }
+                                });
+                            }
+                            else
+                            {
+                                parent.invokeOnComponent(facesContext, targetClientId, topCallback);
+                            }
+                            return;
+                        }
+                    }
+                }
+            }
+
+            if (target != null)
+            {
+                if (nextExpression != null)
+                {
+                    handler.invokeOnComponent(searchExpressionContext, target, nextExpression, topCallback);
+                }
+                else
+                {
+                    topCallback.invokeContextCallback(facesContext, target);
+                }
+
+                return;
+            }
+
+            // At this point if the algorithm hasn't returned and the topExpression does not have any separator char
+            // we need to do the search backward using findComponent.
+            if (target == null 
+                    && searchExpressionContext.getSource() == previous 
+                    && (topExpression.indexOf(separatorChar) == -1) )
+            {
+                UIComponent baseNC = previous.getNamingContainer();
+                if (baseNC != null && baseNC.getParent() != null)
+                {
+                    UIComponent parentNC = getParentNamingContainerUIViewRoot(baseNC.getParent());
+                    while (target == null && parentNC != null)
+                    {
+                        UIComponent parent = parentNC.getParent();
+                        target = parentNC.findComponent(expression);
+                        if (parent != null)
+                        {
+                            parentNC = getParentNamingContainerUIViewRoot(parent);
+                        }
+                        else
+                        {
+                            parentNC = null;
+                        }
+                    }
+                    if (target != null)
+                    {
+                        topCallback.invokeContextCallback(facesContext, target);
+                        return;
+                    }
+                }
+            }
+            
+            if (target == null && nextExpression == null)
+            {
+                return;
+            }
+
+            topCallback.invokeContextCallback(facesContext, previous);
+        }
+        
+    }
+    
+    private static UIComponent getParentNamingContainerUIViewRoot(UIComponent component)
+    {
+        do
+        {
+            if (component instanceof NamingContainer || component instanceof UIViewRoot)
+            {
+                return component;
+            }
+
+            component = component.getParent();
+        } while (component != null);
+        return null;
+    }
+
+    // take the command and resolve it using the chain of responsibility pattern.
+    protected void applyKeyword(SearchExpressionContext searchExpressionContext, UIComponent last,
+                             String command, String remainingExpression, ContextCallback topCallback)
+    {
+        SearchKeywordContext searchContext =
+                new SearchKeywordContext(searchExpressionContext, topCallback, remainingExpression);
+
+        searchExpressionContext.getFacesContext().getApplication()
+                .getSearchKeywordResolver().resolve(searchContext, last, command);
+    }
+
+    @Override
+    public boolean isPassthroughExpression(SearchExpressionContext searchExpressionContext, String topExpression)
+    {
+        if (StringUtils.isBlank(topExpression))
+        {
+            return false;
+        }
+
+        topExpression = topExpression.trim();
+        
+        FacesContext facesContext = searchExpressionContext.getFacesContext();
+
+        //Step 1: find base
+        //  Case ':' (root)
+        char separatorChar = facesContext.getNamingContainerSeparatorChar();
+        if (topExpression.charAt(0) == separatorChar)
+        {
+            // only keywords are passthrough expressions.
+            return false;
+        }
+
+        //Step 2: Once you have a base where you can start, apply an expression
+        if (topExpression.charAt(0) == KEYWORD_PREFIX.charAt(0))
+        {
+            // A keyword means apply a command over the current source using an expression and the result must be
+            // feedback into the algorithm.
+
+            String command = extractKeyword(topExpression, 1, separatorChar);
+            final String remaining =
+                    command.length()+1 < topExpression.length() ?
+                        topExpression.substring(1+command.length()+1) : null;
+
+            final SearchExpressionHandler currentInstance =
+                    facesContext.getApplication().getSearchExpressionHandler();
+
+            // If the keyword is @child, @composite, @form, @namingcontainer, @next, @none, @parent, @previous,
+            // @root, @this ,  all commands change the source to be applied the action
+            boolean passthrough = facesContext.getApplication().getSearchKeywordResolver().isPassthrough(
+                    searchExpressionContext, command);
+
+            if (passthrough)
+            {
+                return remaining != null ?
+                        currentInstance.isPassthroughExpression(searchExpressionContext, remaining) : true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+        else
+        {
+            // Only keywords are valid to be passthrough. If it contains a chain of ids, this can only be resolved
+            // server side, because the tree structure and related clientId logic is only available server side.
+            return false;
+        }
+    }
+
+    @Override
+    public boolean isValidExpression(SearchExpressionContext searchExpressionContext, String topExpression)
+    {
+        if (StringUtils.isBlank(topExpression))
+        {
+            return true;
+        }
+
+        topExpression = topExpression.trim();
+        
+        FacesContext facesContext = searchExpressionContext.getFacesContext();
+        // Command pattern to apply the keyword or command to the base and then invoke the callback
+        boolean isValid = true;
+        //Step 1: find base
+        //  Case ':' (root)
+        char separatorChar = facesContext.getNamingContainerSeparatorChar();
+        if (topExpression.charAt(0) == separatorChar)
+        {
+            return facesContext.getApplication().getSearchExpressionHandler().isValidExpression(
+                    searchExpressionContext, topExpression.substring(1));
+        }
+
+        //Step 2: Once you have a base where you can start, apply an expression
+        if (topExpression.charAt(0) == KEYWORD_PREFIX.charAt(0))
+        {
+            // A keyword means apply a command over the current source using an expression and the result must be
+            // feedback into the algorithm.
+
+            String command = extractKeyword(topExpression, 1, separatorChar);
+            final String remaining =
+                    command.length()+1 < topExpression.length() ?
+                        topExpression.substring(1+command.length()+1) : null;
+
+            final SearchExpressionHandler currentInstance =
+                    facesContext.getApplication().getSearchExpressionHandler();
+
+            // If the keyword is @child, @composite, @form, @namingcontainer, @next, @none, @parent, @previous,
+            // @root, @this ,  all commands change the source to be applied the action
+            isValid = facesContext.getApplication().getSearchKeywordResolver().isResolverForKeyword(
+                    searchExpressionContext, command);
+            if (remaining != null)
+            {
+                if (facesContext.getApplication().getSearchKeywordResolver().isLeaf(
+                    searchExpressionContext, command))
+                {
+                    isValid = false;
+                }
+                return !isValid ? false : currentInstance.isValidExpression(searchExpressionContext, remaining);
+            }
+        }
+        else
+        {
+            //Split expression into tokens and apply loop
+            String nextExpression = null;
+            String expression = null;
+            if (topExpression.indexOf(":@") > 0)
+            {
+                int idx = topExpression.indexOf(":@");
+                nextExpression = topExpression.substring(idx+1);
+                expression = topExpression.substring(0,idx);
+            }
+            else
+            {
+                expression = topExpression;
+            }
+
+            //Check expression
+            for (int i = 0; i < expression.length(); i++)
+            {
+                char c = expression.charAt(i);
+                if (Character.isLetterOrDigit(c) || c == '-' || c == '_' || c == separatorChar)
+                {
+                    //continue
+                }
+                else
+                {
+                    isValid = false;
+                }
+            }
+
+            if (nextExpression != null)
+            {
+                return !isValid ? false : facesContext.getApplication().getSearchExpressionHandler()
+                    .isValidExpression(searchExpressionContext, nextExpression);
+            }
+        }
+        return isValid;
+    }
+
+    private static String extractKeyword(String expression, int startIndex, char separatorChar)
+    {
+        int parenthesesCounter = -1;
+        int count = -1;
+        for (int i = startIndex; i < expression.length(); i++)
+        {
+            char c = expression.charAt(i);
+            if (c == '(')
+            {
+                if (parenthesesCounter == -1)
+                {
+                    parenthesesCounter = 0;
+                }
+                parenthesesCounter++;
+            }
+            if (c == ')')
+            {
+                parenthesesCounter--;
+            }
+            if (parenthesesCounter == 0)
+            {
+                //Close first parentheses
+                count = i+1;
+                break;
+            }
+            if (parenthesesCounter == -1)
+            {
+                if (c == separatorChar)
+                {
+                    count = i;
+                    break;
+                }
+            }
+        }
+        if (count == -1)
+        {
+            return expression.substring(startIndex);
+        }
+        else
+        {
+            return expression.substring(startIndex, count);
+        }
+    }
+
+    @Override
+    public String[] splitExpressions(FacesContext context, String expressions)
+    {
+        // split expressions by blank or comma (and ignore blank and commas inside brackets)
+        String[] splittedExpressions = split(context, expressions, EXPRESSION_SEPARATOR_CHARS);
+        return splittedExpressions;
+    }
+
+    private static String[] split(FacesContext context, String value, char... separators)
+    {
+        if (StringUtils.isBlank(value))
+        {
+            return null;
+        }
+
+        List<String> tokens = new ArrayList<>(5);
+        StringBuilder buffer = SharedStringBuilder.get(context, SB_SPLIT);
+
+        int parenthesesCounter = 0;
+
+        int length = value.length();
+        for (int i = 0; i < length; i++)
+        {
+            char c = value.charAt(i);
+            if (c == '(')
+            {
+                parenthesesCounter++;
+            }
+
+            if (c == ')')
+            {
+                parenthesesCounter--;
+            }
+
+            if (parenthesesCounter == 0)
+            {
+                boolean isSeparator = false;
+                for (char separator : separators)
+                {
+                    if (c == separator)
+                    {
+                        isSeparator = true;
+                    }
+                }
+
+                if (isSeparator)
+                {
+                    // lets add token inside buffer to our tokens
+                    String bufferString = buffer.toString().trim();
+                    if (bufferString.length() > 0)
+                    {
+                        tokens.add(bufferString);
+                    }
+                    // now we need to clear buffer
+                    buffer.delete(0, buffer.length());
+                }
+                else
+                {
+                    buffer.append(c);
+                }
+            }
+            else
+            {
+                buffer.append(c);
+            }
+        }
+
+        // lets not forget about part after the separator
+        tokens.add(buffer.toString());
+
+        return tokens.toArray(new String[tokens.size()]);
+    }
+
+}

--- a/dev/io.openliberty.org.apache.myfaces.4.1/src/org/apache/myfaces/component/search/SearchExpressionHandlerImpl.java
+++ b/dev/io.openliberty.org.apache.myfaces.4.1/src/org/apache/myfaces/component/search/SearchExpressionHandlerImpl.java
@@ -535,7 +535,7 @@ public class SearchExpressionHandlerImpl extends SearchExpressionHandler
             // we need to do the search backward using findComponent.
             if (target == null 
                     && searchExpressionContext.getSource() == previous 
-                    && (topExpression.indexOf(separatorChar) == -1) )
+                    && (topExpression.indexOf(separatorChar + "@") == -1) ) // updated for MYFACES-4695
             {
                 UIComponent baseNC = previous.getNamingContainer();
                 if (baseNC != null && baseNC.getParent() != null)

